examples_apps/run_gate_synthesis.py

```diff
@@ -0,0 +1,453 @@
+with prog.context as q:
+    Ket(in_ket) | q
+    for k in range(depth):
+        layer(k) | q[0]
+
+######################################################################
+# Performing the optimization
+# ---------------------------
+#
+# :math:`\newcommand{ket}[1]{\left|#1\right\rangle}`
+# With the Strawberry Fields TensorFlow backend calculating the
+# resulting state of the circuit symbolically, we can use TensorFlow to
+# optimize the gate parameters to minimize the cost function we specify.
+# With gate synthesis, we minimize the overlaps in the Fock basis
+# between the target and learnt unitaries via the following cost function:
+#
+# .. math::
+#
+#     C(\vec{\theta}) = \frac{1}{d}\sum_{i=0}^{d-1} \left| \langle i \mid
+#     V^\dagger U(\vec{\theta})\mid i\rangle - 1\right|
+#
+# where :math:`V` is the target unitary, :math:`U(\vec{\theta})` is the
+# learnt unitary, and :math:`d` is the gate cutoff. Note that this is a
+# generalization of state preparation to more than one input-output
+# relation.
+#
+# For our target unitary, let's use Strawberry Fields to generate a 4x4
+# random unitary:
+#
+
+from strawberryfields.utils import random_interferometer
+target_unitary = np.identity(cutoff, dtype=np.complex128)
+target_unitary[:gate_cutoff, :gate_cutoff] = random_interferometer(4)
+print(target_unitary)
+
+
+######################################################################
+# This matches the gate cutoff of :math:`d=4` that we chose above when
+# defining our hyperparameters.
+#
+# Now, we instantiate the Strawberry Fields TensorFlow backend:
+
+eng = sf.Engine('tf', backend_options={"cutoff_dim": cutoff, "batch_size": gate_cutoff})
+
+######################################################################
+# Here, we use the ``batch_size`` argument to perform the optimization in
+# parallel - each batch calculates the variational quantum circuit acting
+# on a different input Fock state: :math:`U(\vec{\theta}) | n\rangle`.
+#
+
+in_state = np.arange(gate_cutoff)
+
+# extract action of the target unitary acting on
+# the allowed input fock states.
+target_kets = np.array([target_unitary[:, i] for i in in_state])
+target_kets = tf.constant(target_kets, dtype=tf.complex64)
+
+######################################################################
+# Using this target unitary, we define the cost function we would like to
+# minimize. We must use TensorFlow functions to manipulate this data, as well
+# as a ``GradientTape`` to keep track of the corresponding gradients!
+#
+def cost(weights):
+    # Create a dictionary mapping from the names of the Strawberry Fields
+    # free parameters to the TensorFlow weight values.
+    mapping = {p.name: w for p, w in zip(sf_params.flatten(), tf.reshape(weights, [-1]))}
+
+    # Run engine
+    state = eng.run(prog, args=mapping).state
+
+    # Extract the statevector
+    ket = state.ket()
+
+    # # overlaps
+    overlaps = tf.math.real(tf.einsum('bi,bi->b', tf.math.conj(target_kets), ket))
```

Comment:
Note the presence of einsum here, and that its gradient is incorrect for complex numbers.
See https://github.com/XanaduAI/strawberryfields/pull/377.

Reply:
Thanks @trbromley !

Reply:
Now that #377 has been merged in, this call to `tf.einsum` will automatically call `_einsum_v1`, due to a quirk of how Python imports work :)